packages/live-share/src/internals/utils.ts

```diff
@@ -41,46 +45,120 @@ export function waitForDelay(delay: number): Promise<void> {
+                    resolve(validated.response);
+                    break;
+                } else if (retries >= retrySchedule.length) {
+                    reject(new Error("waitForResult: invalid response"));
+                    break;
+                }
+            } catch (error: unknown) {
+                if (retries >= retrySchedule.length) {
+                    reject(fnTimeout(error));
+                    break;
+                }
+                // Check if this error is something that should cause us to skip the retry schedule
+                if (!!fnRequestError && !(error instanceof TimeoutError)) {
+                    const rejectNowError = fnRequestError(error);
+                    if (rejectNowError !== null) {
+                        reject(rejectNowError);
+                        break;
+                    }
+                }
             }
+            
+            await waitForDelay(retrySchedule[retries++]);
         }
     });
 }
 
-// BUGBUG: Workaround for Teams Client not rejecting errors :(
+/**
+ * BUGBUG: Workaround for Teams Client not rejecting errors :(
+ * @hidden
+ */
 function timeoutRequest<TResult>(
-    fnRequest: () => Promise<TResult | undefined>,
+    fnRequest: () => Promise<TResult>,
     timeout: number
-): Promise<TResult | undefined> {
-    return new Promise<TResult | undefined>(async (resolve) => {
+): Promise<TResult> {
+    return new Promise<TResult>(async (resolve, reject) => {
         const hTimer = setTimeout(() => {
-            resolve(undefined);
+            reject(new TimeoutError());
         }, timeout);
-        const result = await fnRequest();
+        try {
+            const result = await fnRequest();
+            resolve(result);
+        } catch (error: unknown) {
+            reject(error);
+        }
         clearTimeout(hTimer);
-        resolve(result);
+    });
+}
+
+/**
+ * Dynamically import InsecureTokenProvider class, in case developer does not yet have "@fluidframework/test-client-utils",
+ * since don't want to require that they include it in package.json.
+ * @hidden
+ */
+export async function getInsecureTokenProvider(): Promise<ITokenProvider> {
+    try {
+        const { InsecureTokenProvider } =
+            await require("@fluidframework/test-client-utils");
+        const userIdParam = new URL(window.location.href)?.searchParams?.get("userId");
+        const tokenProvider = new InsecureTokenProvider("", {
+            id: userIdParam ?? uuid(),
+            name: "Test User",
+        });
+        return tokenProvider as ITokenProvider;
+    } catch {
+        throw new Error(
+            "@microsoft/live-share: when using 'local' connection type, you must have @fluidframework/test-client-utils installed"
+        );
+    }
+}
+
+/**
+ * @hidden
+ * Waits until connected and gets the most recent clientId
+ * @returns clientId
+ */
+export function waitUntilConnected(runtime: IRuntimeSignaler): Promise<string> {
```

Comment:
update usage in LiveMediaSessionCoordinator add LiveEventScope

Reply:
hmm currently this isn't publicly exported. guess we can export and hide?

what's the comment on "add LiveEventScope"?